packages/toolpad-app/src/toolpad/AppEditor/index.tsx

```diff
@@ -43,48 +42,73 @@ interface FileEditorProps {
 }
 
 function FileEditor({ appId }: FileEditorProps) {
-  const { dom, currentView } = useDom();
-
-  const app = appDom.getApp(dom);
-  const { pages = [] } = appDom.getChildNodes(dom, app);
+  const { currentView } = useDom();
 
+  const location = useLocation();
   const navigate = useNavigate();
 
-  const firstPage = pages.length > 0 ? pages[0] : null;
-
   React.useEffect(() => {
```

Comment:
I believe this could be a bit DRYer if written as:

```tsx
function getPathnameForView (appId, view) {
  switch (view.kind) {
    case 'page': return `/app/${appId}/pages/${view.nodeId}`;
    // ...
  }
}

React.useEffect(() => {
  const desiredPath = getPathnameForView(currentView)
  if (desiredPath !== location.pathname) {
    navigate({  pathname: desiredPath }, { replace: true });
  }
}, [appId, currentView, location.pathname]
```